src/main/scala/com/fulcrumgenomics/sv/tools/AggregateSvPileup.scala

```diff
@@ -0,0 +1,281 @@
+        map.put(pileup1, pileup2 :: map(pileup1))
+        map.put(pileup2, pileup1 :: map(pileup2))
+      }
+    }
+    map.toMap
+  }
+
+  /** Returns true if the two breakpoints can be clustered, i.e., they are not the same breakpoint (have different
+   * IDs), they have the same left and right chromosomes and strand, and the distance between their left and right
+   * coordinates is below the threshold (inclusive) */
+  private def isCluster(bp1: BreakpointPileup, bp2: BreakpointPileup, maxDist: Int): Boolean = {
+    bp1.id != bp2.id &&
+      BreakpointContigOrientation(bp1) == BreakpointContigOrientation(bp2) &&
+      Math.abs(bp1.left_pos - bp2.left_pos) <= maxDist &&
+      Math.abs(bp1.right_pos - bp2.right_pos) <= maxDist
+  }
+
+  /**
+   * Converts a set of pairwise neighbor relationships to a set of clusters, where each cluster consists of breakpoints
+   * that can be joined into a connected path via pairwise neighbor relationships.
+   * @param pileupToNeighbors Map of pileup to its set of neighbors
+   * @param existingClusters An existing set of clusters to add to
+   * @return The set of pileup clusters
+   */
+  @tailrec
+  def toClusters(pileupToNeighbors: Map[BreakpointPileup, PileupGroup],
+                 existingClusters: Seq[PileupGroup] = Seq()): Seq[PileupGroup] = {
+    if (pileupToNeighbors.isEmpty) {
+      existingClusters
+    } else {
+      val (pileupGroup, remaining) = extractClusterFor(pileupToNeighbors.keys.head, pileupToNeighbors)
+      toClusters(remaining, existingClusters :+ pileupGroup)
+    }
+  }
+
+  /**
+   * Extracts the complete connected cluster containing the given pileup from the map of pileups to their neighbor sets,
+   * and returns the cluster along with the reduced map of pileup to neighbor set such that no remaining pileups in the
+   * map belong in a cluster with the returned cluster.
+   *
+   * Implementation note: the implementation is not tail-recursive, but in practice the recursion depth is small.
+   *
+   * @param pileup The pileup representative of the cluster that will be returned
+   * @param pileupToNeighbors Map of pileup to its set of neighbors
+   * @param existingCluster An existing cluster to add to
+   * @return (1) The complete cluster for the given pileup; (2) Reduced version of `pileupToNeighbors` with all pileups
+   *         in the returned cluster having been removed from both keys and values.
+   */
+  def extractClusterFor(pileup: BreakpointPileup, pileupToNeighbors: Map[BreakpointPileup, PileupGroup],
+                        existingCluster: PileupGroup = Seq()): (PileupGroup, Map[BreakpointPileup, PileupGroup]) = {
+
+    // Identify the immediate neighbors of the given pileup, and start a cluster containing them
+    val neighbors: PileupGroup = pileupToNeighbors.getOrElse(pileup, Seq.empty)
+    val currCluster: PileupGroup = neighbors :+ pileup
+
+    // Remove the given pileup and its immediate neighbors from `pileupToNeighbors`
```

Comment:
I don't think the `neighbors` are removed yet
```suggestion
    // Remove the given pileup from `pileupToNeighbors`
```